Fix batch_norm dtype mismatch on MPS (fixes #154887)

roei-shlezinger opened the pull request:
This PR fixes a dtype mismatch in MPS batch_norm where float16 inputs and float32 running stats caused MPSGraph to fail during type inference.
The fix adds a small dtype alignment step (casting mean/var/weight/bias to the input dtype when needed).
No functional or numerical changes-only dtype consistency before calling normalizationWithTensor.

Fixes #154887

roei-shlezinger committed:
Fix MPS batch norm dtype mismatch for running stats

kurtamohler commented on test/test_mps.py:
```diff
@@ -1831,6 +1831,28 @@ def test_batch_norm_backward_weight_bias_gradients(self):
 
+    def test_batch_norm_mixed_dtypes(self):
+        # See issue: https://github.com/pytorch/pytorch/issues/154887#issue-3111485538
+        input_mps = torch.rand((2, 3, 4), dtype=torch.float16, device="mps")
```
Looks like your change will enable any `dtype_a` -> `dtype_b` cast, not just float16 -> float32, so we should probably test all of the supported cases, using `@parametrize`.

Also, I think the CPU impl does not support all cases and raises an error instead, so MPS should probably error in the same cases

kurtamohler reviewed (commented):
Thanks for the fix, just a few suggestions

roei-shlezinger committed:
mps: align batch_norm casts/dtype checks with CPU and expand test

- use castMPSTensor for batch_norm input params instead of custom lambda
- enforce CPU mixed-dtype rules (is_mixed_type/check_mixed_data_type) before building the MPS graph
- expand batch_norm mixed-dtype test with allowed/blocked combos; clean up CPU clones

kurtamohler reviewed (approved):
Just a few nits but it looks good to me.

I think the `check_mixed_data_type` oddity is ok. It would be good to fix its error message to be device-agnostic and move it to a more common place than `ATen/native/cpu`, but that should be done in a separate PR

roei-shlezinger committed:
mps: simplify batch norm mixed dtype test

kurtamohler commented:
@pytorchbot merge -r
